pkgs/development/tools/build-managers/meson/default.nix

```diff
@@ -1,7 +1,6 @@
-{ lib, python3Packages, stdenv, targetPlatform, writeTextDir, substituteAll }: let
-  targetPrefix = lib.optionalString stdenv.isCross
-                   (targetPlatform.config + "-");
-in python3Packages.buildPythonApplication rec {
+{ lib, python3Packages, stdenv, targetPlatform, writeTextDir, substituteAll }:
```

Comment:
Why does meson care about `targetPlatform`? Sounds like a bug (an existing one though).

Reply:
Meson is a bit confused with cross :(. I was hoping to fix it, starting with https://github.com/mesonbuild/meson/pull/3689 which I thought would be uncontroversial. But it was, and then I think I burnt some credibility in the process.

Reply:
I mean a bug in the Nix expression. This can't be right?

````
 60     [host_machine]
 61     system = '${targetPlatform.parsed.kernel.name}'
 62     cpu_family = '${targetPlatform.parsed.cpu.family}'
 63     cpu = '${targetPlatform.parsed.cpu.name}'
 64     endian = ${if targetPlatform.isLittleEndian then "'little'" else "'big'"}
````
The only place where you care about `targetPlatform` is when building a cross compiler. I don't see why meson would ever apply here.

Reply:
Also I don't understand what "Meson is a bit confused with cross" means. Based on my reading, it has mostly the same semantics and names for everything as autoconf?

Reply:
It's not *that* bad. I fixed some terminology in https://github.com/mesonbuild/meson/blob/master/docs/markdown/Cross-compilation.md without issue (i.e. the thing itself did the mostly right thing) for example.

The concrete issue is that pops up here is Meson handles compiler env vars incorrectly.

With autoconf, `CC_FOR_BUILD` (or long ago + still in Linux `HOST_CC`) targets the build platform, `CC` targets the host platform, and `CC_FOR_TARGET` targets the target platform. The `_FOR_*` ones are defaulted in the native case so the build system writer doesn't need any special casing. With Meson, `CC` always targets the native platform, so sometimes it's for build and host, and sometimes it's  for build.

What's going in Nix is a) we need to wip out our autoconf-style `CC` (and other variables, which isn't done yet) and b) we are making a cross file baked into meson, which is also sketchy cause we don't know what compilers we'll have in actuality. Since meson is making the file, and already a `nativeBuildInput` it's only a little stretch to pretend it's a cross compile and use `targetPlatform` according too for the build hook bug yuck.

This reflects a larger issue that meson's *implementation* constantly thinks "native vs foreign" rather than "build vs host vs target" which leads to tons of needless and hazardous conditional code.

I suppose I opened https://github.com/mesonbuild/meson/issues/3282 before I knew it used `CC` incorrectly in the cross case, I suppose I should open another making it use the env vars like autoconf.

Reply:
> With Meson, CC always targets the native platform, so sometimes it's for build and host, and sometimes it's for build.

Right, if setting CC overrides the native compiler even when a cross file is present, that will be painful. Having them support `CC_FOR_BUILD` would be useful indeed.

> we are making a cross file baked into meson, which is also sketchy cause we don't know what compilers we'll have in actuality.

You can create the cross file in the setup hook at build-time. I presume it would be needed anyway if you ever need to tell meson about say, a cross Fortran compiler and not have meson depend on gfortran all the time.

> This reflects a larger issue that meson's implementation constantly thinks "native vs foreign" rather than "build vs host vs target" which leads to tons of needless and hazardous conditional code.

Any concrete example?

Reply:
> Having them support CC_FOR_BUILD would be useful indeed.

Glad you agree. Made https://github.com/mesonbuild/meson/issues/3969 for `CC`/`CC_FOR_BUILD`.

> You can create the cross file in the setup hook at build-time.

Yeah agree making on the fly (probably via setup-hook) is a more reliable way to get this done. Of if we deem Meson important enough, we can make it part of `-wrapper`. (If I do the big refactors I want, probably at the end up the month, those should be a lot simpler freeing up some complexity budget.)

> Any concrete example?

grep for `is_cross` I recall. I had some local changes I could dredge up too.

Reply:
(In https://github.com/mesonbuild/meson/pull/3689#issuecomment-410082269, I raise trying different plans of attack that people might be more on board with.)